File: Chat  - exercise-4/find_and_print_files.rb
# Searches for files by name or prints every regular file under the script's directory.
# The --all option enables printing the complete folder tree.
# frozen_string_literal: true

require "find"

def print_files(paths)
  paths.sort.each do |path|
    puts "\n--- #{path} ---"
    contents = File.read(path)
    print contents
    puts unless contents.end_with?("\n")
  end
end

def print_all_files(directory)
  paths = []

  Find.find(directory) do |path|
    paths << path if File.file?(path)
  end

  print_files(paths)
end

file_name = ARGV[0]

if file_name == "--all"
  print_all_files(__dir__)
  exit 0
end

if file_name.nil? || file_name.empty?
  warn "Usage: ruby find_and_print_files.rb FILE_NAME|--all"
  exit 1
end

matches = []

Find.find(__dir__) do |path|
  next unless File.file?(path)
  next unless File.basename(path) == file_name

  matches << path
end

if matches.empty?
  puts "No files named #{file_name.inspect} found in #{__dir__}."
  exit 0
end

print_files(matches)

File: Chat  - exercise-4/find_and_print_files.py
#!/usr/bin/env python3

import os
import sys


def main() -> int:
    if len(sys.argv) < 2 or not sys.argv[1]:
        print("Usage: python find_and_print_files.py FILE_NAME", file=sys.stderr)
        return 1

    file_name = sys.argv[1]
    script_directory = os.path.dirname(os.path.abspath(__file__))
    matches = []

    for directory_path, _, file_names in os.walk(script_directory):
        for name in file_names:
            path = os.path.join(directory_path, name)
            if name == file_name and os.path.isfile(path):
                matches.append(path)

    if not matches:
        print(f"No files named {file_name!r} found in {script_directory}.")
        return 0

    for path in sorted(matches):
        print(f"\n--- {path} ---")
        with open(path, encoding="utf-8") as file:
            contents = file.read()
        print(contents, end="" if contents.endswith("\n") else "\n")

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
